### PR TITLE
Add category: "Economics" to PV tree, add a few new specs under economics

### PR DESCRIPTION
**Pull request by qlj-lijuan**
- fixed issue [#273](https://github.com/datacommonsorg/website/issues/273). The economics category is missing because it's named as "Economic".
- add a few new specs related to the statsVars:
https://datacommons.org/browser/GrowthRate_Amount_EconomicActivity_GrossDomesticProduction
https://datacommons.org/browser/Amount_EconomicActivity_GrossDomesticProduction_Nominal_PerCapita
https://datacommons.org/browser/Amount_EconomicActivity_GrossDomesticProduction_Nominal
https://datacommons.org/browser/Amount_EconomicActivity_GrossDomesticProduction_RealValue
- Currently, "measurementQualifier" and "measurementDenominator" are treated as constraint properties. 
- The new specs shown in the PV tree as:
image
- We have some old specs under the "Economics" category, which seems to fit better under employment
image

### CODE REVIEW

**Comment by beets** [comment]
Thanks for this update! I wonder if there is a way to order the list under economics. I would imagine that measures of GDP are more important than establishment counts. It would be helpful to see the expanded list in the screenshot.

@jeffreyoldham for suggestions on how we might best rank these economics stats vars.

**Comment by qlj-lijuan** [comment]
> Thanks for this update! I wonder if there is a way to order the list under economics. I would imagine that measures of GDP are more important than establishment counts. It would be helpful to see the expanded list in the screenshot.
> 
> @jeffreyoldham for suggestions on how we might best rank these economics stats vars.

yeah. We can sort them.

**Comment by qlj-lijuan** [comment]
Add Obs_props to support multiple nodes under one pop_obs_spec
Add additional layer when there're multiple nodes under the same pop_obs_spec, but with different stats_type, measurement denominator etc.
image

File: static/data/hierarchy_statsvar.json [hunk 1]
```diff
@@ -162425,13 +162419,249 @@
...
+      ],
+      "l": "Unclassified",
+      "t": "v",
+      "e": "NAICS/99",
+      "c": 1
+     }
+    ]
+   },
+   {
+    "l": "Measurement Qualifier",
```
[the remark this answers is not in the document]
Comment (qlj-lijuan) [reply static/data/hierarchy_statsvar.json:162623]: This is an old version. The new labels are in the screenshot above.

File: tools/pv_tree_generator/stat_config.proto [hunk 2]
```diff
@@ -44,6 +52,7 @@ message PopObsSpec {
   optional string name = 6;
   // Verticals of this spec.
   repeated string vertical = 7;
+  repeated ObsProp obs_props = 8;
```
Comment (shifucun) [inline tools/pv_tree_generator/stat_config.proto:55]: With this, can remove line 42 and 44?

This also means we can treat the current "count", "median age", "median income" in a more general way as they are the same population with different ObsProp.

File: tools/pv_tree_generator/util.py [hunk 3]
```diff
@@ -61,24 +74,26 @@ def match_ui_node(self, ui_node):
 class UiNode:
     """Class holds the data of a node in the pv menu tree. """
 
-    def __init__(self, pop_obs_spec, pv, is_prop, prop=None):
+    def __init__(self, pop_obs_spec, obs_props, pv, is_prop, prop=None):
```
Comment (shifucun) [inline tools/pv_tree_generator/util.py:77]: the second argument pop_obs_spec should already have obs_props?

File: tools/pv_tree_generator/util.py [hunk 4]
```diff
@@ -203,24 +228,26 @@ def _read_stat_var():
...
+        se = {}
         if 'crimeType' in prop_val:
             v = prop_val.get('crimeType', '')
             if v in ['AggravatedAssault', 'ForcibleRape', 'Robbery',
                      'MurderAndNonNegligentManslaughter']:
                 se = {'crimeType': 'ViolentCrime'}
             elif v in ['MotorVehicleTheft', 'LarcenyTheft', 'Burglary']:
                 se = {'crimeType': 'PropertyCrime'}
+        # create the statsVar object
         sv = StatVar(sv_dict["populationType"], sv_dict["measuredProperty"],
```
Comment (shifucun) [inline tools/pv_tree_generator/util.py:249]: nit: new line for each argument when constructing StatVar

File: tools/pv_tree_generator/build_tree.py [hunk 5]
```diff
@@ -47,59 +47,90 @@ def build_tree_recursive(pos, level, pop_obs_spec, stat_vars,
...
-                value_blob['c'] = len(value_blob['sv_set'])
-            else:
-                # if this is a duplicated statsVar
-                # append the statsVar to the right node, increase the count by 1
-                for child in result['cd']:
-                    if child['e'] == sv.pv[property_diff]:
-                        child['sv'].append(sv.dcid)
+    newLevel = (len(pos.obs_props) > 1) # create a new level if there're multiple obs_props
+    child_val_nodes = {}
+    for obs in pos.obs_props:
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:55]: i am not sure if the obs is the outer  for loop. I think the stats menu level should be purely based on population properties like "age", "gender". The spec with multiple ObsProps would have a few leaf nodes (with check box) but the children expansion is still based on cprops.

Can take Person: "median age", "median income", "[age=xx]", "[gender=xx]" as an example. Demographs will have a "median age", "median income" as child value nodes, and age, gender as child property nodes.

File: tools/pv_tree_generator/build_tree.py [hunk 6]
```diff
@@ -47,59 +47,90 @@ def build_tree_recursive(pos, level, pop_obs_spec, stat_vars,
...
+        key = tuple([pos.pop_type])+ obs.key + pos.prop_all
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:56]: the for loop here and child_val_nodes are a bit confusing. A better way would be handle all the stats_var selection in a class.

For example, you can have brand new class called StatsVarStore, in the initialization, it will

- read all stats var
- filter duplicate stats vars
- make a private member variable indexed by the "key" here.
- expose a function like get_statsvar(pop_obs_spec, parent_pv), and return a list of stats_vars. This way can save the current line 57, 58 maybe 62, 63

File: tools/pv_tree_generator/util.py (hunk 4 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/util.py]: done

File: tools/pv_tree_generator/stat_config.proto [hunk 7]
```diff
@@ -21,6 +21,14 @@ package datacommons.proto;
 // This proto contains the specification of a PopObs instance. It contains
```
[the remark this answers is not in the document]
Comment (qlj-lijuan) [reply tools/pv_tree_generator/stat_config.proto]: done

File: tools/pv_tree_generator/stat_config.proto [hunk 8]
```diff
@@ -21,6 +21,14 @@ package datacommons.proto;
...
 // the intended population type, measured property, stats type, constraining
 // property and hidden property-value pairs.
+message ObsProp {
+  optional string stat_type =1;
```
[the remark this answers is not in the document]
Comment (qlj-lijuan) [reply tools/pv_tree_generator/stat_config.proto]: done.

File: tools/pv_tree_generator/stat_config.proto (hunk 2 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/stat_config.proto:60]: kept it for supporting the old specs, otherwise we would need to revise all the old spec text protos

File: static/data/statsvar_path.json [hunk 9]
```diff
@@ -2993,14 +2993,14 @@
   5,
   1,
   0,
-  5
+  4
```
Comment (qlj-lijuan) [inline static/data/statsvar_path.json:2996]: This is the right order. The old one is not right.

File: static/data/statsvar_path.json [hunk 10]
```diff
@@ -16650,7 +16636,7 @@
   0,
   2
  ],
- "Count_CriminalActivities_Burglary": [
+ "Count_CriminalActivities_MotorVehicleTheft": [
```
Comment (qlj-lijuan) [inline static/data/statsvar_path.json:16639]: The crime types are not sorted, the random order is expected.

File: tools/pv_tree_generator/stat_config.proto (hunk 2 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/stat_config.proto:55]: removed mprop and stats_type, revised the proto text

File: tools/pv_tree_generator/stat_config.proto [hunk 11]
```diff
@@ -21,6 +21,14 @@ package datacommons.proto;
...
+  optional string mprop=2;
+  optional string mqual=3;
+  optional string mdenom=4;
+  optional string sfactor=5;
```
[the remark this answers is not in the document]
Comment (qlj-lijuan) [reply tools/pv_tree_generator/stat_config.proto]: removed sfactor. Scaling factor and unit are moved to observations

File: static/js/place/place_template.tsx [hunk 12]
```diff
@@ -751,9 +751,6 @@ class Chart extends Component<ChartPropType, ChartStateType> {
           for (const dataGroup of dataGroups) {
             dataGroup.label = STATS_VAR_TEXT[dataGroup.label];
           }
-          this.setState({
```
Comment (shifucun) [inline static/js/place/place_template.tsx:754]: why is this removed?

File: tools/pv_tree_generator/build_tree.py [hunk 13]
```diff
@@ -1,243 +1,329 @@
-# Copyright 2020 Google LLC
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:1]: please add back license header

File: tools/pv_tree_generator/build_tree.py [hunk 14]
```diff
@@ -1,243 +1,329 @@
...
+from typing import Dict, List, Tuple
+from util import PopObsSpec, StatsVar, read_pop_obs_spec, read_stat_var
 import collections
-import util
 import text_format
+from constants import VERTICALS
+import copy
+
+
+class Root:
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:9]: if there is no class variable, then maybe no need to create a class? can just have function get_root_children() and get_root_name()

File: tools/pv_tree_generator/build_tree.py [hunk 15]
```diff
@@ -1,243 +1,329 @@
...
+        }
+        return result
+
+
+class ValueNode:
+    """ Represent a value node in the PV tree
+        The value node can be
+            - a real value node representing a statsVar
+            - the node of a vertical, such as the node "Demographics"
+            - a psudo value node of super enums, such as the node
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:145]: psudo -> pseudo

File: tools/pv_tree_generator/build_tree.py [hunk 16]
```diff
@@ -1,243 +1,329 @@
...
 
-    result['cd'] = text_format.filter_and_sort(property_diff,
-                                               result['cd'], False)
 
-    # update the count
-    if result['cd']:
-        for child in result['cd']:
-            result['sv_set'] |= child['sv_set']
-            del child['sv_set']
+class valueLeaf:
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:214]: ValueLeaf. class name should be capitalized.

File: tools/pv_tree_generator/build_tree.py [hunk 17]
```diff
@@ -1,243 +1,329 @@
...
-                # if this is the first statsvar of this node, build the node
-                # and corresponding child nodes
-                child_prop_vals.append(sv.pv[property_diff])
-                value_ui_pv = collections.OrderedDict()
-                for prop, val in parent_pv.items():
-                    value_ui_pv[prop] = val
-                value_ui_pv[property_diff] = sv.pv[property_diff]
-                value_ui_node = util.UiNode(
-                    pos, value_ui_pv, False, property_diff)
+    def __init__(self, value: str, leafs: List['valueLeaf'],
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:149]: should this be List[valueLeaf] (and actually better to be List[ValueLeaf]?

File: tools/pv_tree_generator/build_tree.py [hunk 18]
```diff
@@ -1,243 +1,329 @@
...
+        self.dcids = dcids
+        self.addLevel = addLevel
+        self.pop_type = pos.pop_type
 
-    result['c'] = len(result['sv_set'])
-    result = group_super_enum(result)
-    return result
 
+def build_tree(max_level):
+    global POP_OBS_SPEC
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:223]: global is not recommended. Can put them in a separate module if you want to share it across other modules.

File: tools/pv_tree_generator/build_tree.py [hunk 19]
```diff
@@ -1,243 +1,329 @@
...
-    for pv0 in root['cd']:
-        if 'sv_set' in pv0:
-            pv0['c'] += len(pv0['sv_set'])
-            del pv0['sv_set']
-    root['c'] += len(root['sv_set'])
-    del root['sv_set']
-    statsvar_path = {}
-    return traverseTree(root, [vertical_idx], statsvar_path)
+
+def build_tree_recursive(node: PropertyNode, max_level):
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:249]: Please add class comments

File: tools/pv_tree_generator/build_tree.py [hunk 20]
```diff
@@ -1,243 +1,329 @@
...
+                # build the next level
+                prop_blob = build_tree_recursive(prop, max_level)
+                if (prop_blob['cd']):
+                    value_blob['cd'].append(prop_blob)
+    result['cd'] = text_format.filter_and_sort(node.prop, result['cd'], False)
+    return result
 
 
-def traverseTree(root, path, statsvar_path):
+def post_process(root, path, statsvar_path):
```
Comment (shifucun) [inline tools/pv_tree_generator/build_tree.py:267]: please add function doc string on what processing is in here

File: static/js/place/place_template.tsx (hunk 12 above)
Comment (qlj-lijuan) [reply static/js/place/place_template.tsx]: Oops. It was removed by accident. Added it back.

File: tools/pv_tree_generator/build_tree.py (hunk 13 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: Done

File: tools/pv_tree_generator/build_tree.py (hunk 14 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: changed to functions.

File: tools/pv_tree_generator/build_tree.py (hunk 15 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: corrected typo

File: tools/pv_tree_generator/build_tree.py (hunk 16 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: Done.

File: tools/pv_tree_generator/build_tree.py (hunk 17 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: Changed to List[valueLeaf].
FYI, I use List['ValueNode'] in class PropertyNode, because to use List[ValueNode], class ValueNode has be defined before PropertyNode.

File: tools/pv_tree_generator/build_tree.py (hunk 18 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py]: Removed global. Pass it to class PropertyNode and VariableNode as arguments, removed the argument 'vertical' in PropertyNode with this change.

File: tools/pv_tree_generator/build_tree.py (hunk 19 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py:276]: Done.

File: tools/pv_tree_generator/build_tree.py (hunk 20 above)
Comment (qlj-lijuan) [reply tools/pv_tree_generator/build_tree.py:296]: Done.
